### PR TITLE
Jebohm extract

### PR DESCRIPTION
**Pull request by jebohm**

### CODE REVIEW

File: tests/data/karelian_nlp_smoke_test_preprocessed.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+###C: {'entry_num': 0}
```
Comment (lumikani) [inline tests/data/karelian_nlp_smoke_test_preprocessed.txt:1]: I don't think this file should be in this commit :sweat_smile:  This file appears temporarily during unit/integration tests for the NLP setup (fin-dep-parser) but since those tests are broken, I guess it does not get cleaned up :thinking:

File: extractors/bookseries/karelians/extractors/religion_extractor.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import core.utils.regex_utils as regexUtils
+from core.pipeline_construction.base_extractor import BaseExtractor
+
+
+class ReligionExtractor(BaseExtractor):
+    extraction_key = 'religion'
+
+    def __init__(self, cursor_location_depends_on=None, options=None):
+        super(ReligionExtractor, self).__init__(cursor_location_depends_on, options)
+        self.RELIGION_PATTERN = r"(?P<ortodoksi>ortodoksi)"
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/religion_extractor.py]: Right now this is extractor is called a `ReligionExtractor` but what it really extracts is just whether the string `"ortodoksi"` appears in their entry. Either the extractor should be renamed to be more accurate, or the implementation needs to be changed to actually extract the person's religion.

File: extractors/bookseries/karelians/extractors/tests/test_religion_extractor.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    def should_correctly_extract_religion(self, extractor):
+        self._verify_religion([
+            ('Rouva X kuuluu ortodoksisen kirkon laulu-kuoroon ja Karjalaseuraan.', True)
+        ], extractor)
+
+    def should_not_extract_religion(self, extractor):
+        self._verify_religion([
+            ('Perheellä on myös maatila, jota he yhteisesti hoitavat. Puolisoiden yhteinen harrastus on seurakuntatyö', False)
+        ], extractor)
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/tests/test_religion_extractor.py]: Missing a newline at the end of the file :thinking:  I will try to set up some sort of linter and/or formatter with a commit hook or something like that, so that things like this don't happen.

File: extractors/bookseries/karelians/extractors/forest_extractor.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import re
+
+import core.utils.regex_utils as regexUtils
+from core.pipeline_construction.base_extractor import BaseExtractor
+
+
+class ForestExtractor(BaseExtractor):
+    extraction_key = 'metsää'
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/forest_extractor.py]: We should stick to English with all of the data output by `kaira` and also in the code itself (eg. names of extractors). Also, most of them are already in English, and maintain consistency is nice. Translating some of the existing keys to English would be nice.

Maybe something like `forestArea` or `forestAreaOwned` or something.. Also there's a farm area extractor, maybe check which key that uses. Would be nice to maintain consistency between the extraction keys :)

File: extractors/bookseries/karelians/extractors/forest_extractor.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import core.utils.regex_utils as regexUtils
+from core.pipeline_construction.base_extractor import BaseExtractor
+
+
+class ForestExtractor(BaseExtractor):
+    extraction_key = 'metsää'
+
+    def __init__(self, cursor_location_depends_on=None, options=None):
+        super(ForestExtractor, self).__init__(cursor_location_depends_on, options)
+        self.FOREST_PATTERN = r'(?:([0-9][0-9]) ha metsää){s<=1}|(?<!H)(a metsää){s<=1}|(metsää [0-9][0-9] ha){s<=1}'
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/forest_extractor.py]: `\d` is the same as writing `[0-9]`, FYI

Comment (lumikani) [reply extractors/bookseries/karelians/extractors/forest_extractor.py]:
```suggestion
        self.FOREST_PATTERN = r'(?:([0-9][0-9])\sha\smetsää){s<=1}|(?<!H)(a\smetsää){s<=1}|(metsää\s[0-9][0-9]\sha){s<=1}'

```
The OCR software that has gone through the scanned books sometimes produces typos, as we discussed. Those typos can even concern `' '` characters (spaces). It's safer to try to match those with `\s`

Comment (lumikani) [reply extractors/bookseries/karelians/extractors/forest_extractor.py]: What's with this negative look~~ahead~~behind? `(?<!H)(a\smetsää){s<=1}` It's trying to capture "a metsää", allowing for one typo, and not allowed to be preceded by the character H? Interesting case. Is it the same as "ha metsää" but using a different unit?

File: extractors/bookseries/karelians/extractors/forest_extractor.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        self.FOREST_OPTIONS = (re.UNICODE | re.IGNORECASE)
+
+    def _extract(self, entry, extraction_results, extraction_metadata):
+        own_forest = self._find_forest(entry["text"])
+        return self._add_to_extraction_results(own_forest, extraction_results, extraction_metadata)
+
+    def _find_forest(self, text):
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/forest_extractor.py]: Maybe `_find_forest_area`? Or `_find_if_owns_forest`? Wait, what was the spec actually, what was this extractor supposed to extract? Looking at the `_find_forest` function implementation, it looks like currently this extractor just extracts a `True`/`False` depending on whether ownership of forests is mentioned. Was that what was supposed to be extracted? Or was the area supposed to be extracted?

File: extractors/bookseries/karelians/extractors/forest_extractor.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+class ForestExtractor(BaseExtractor):
+    extraction_key = 'metsää'
+
+    def __init__(self, cursor_location_depends_on=None, options=None):
+        super(ForestExtractor, self).__init__(cursor_location_depends_on, options)
+        self.FOREST_PATTERN = r'(?:([0-9][0-9]) ha metsää){s<=1}|(?<!H)(a metsää){s<=1}|(metsää [0-9][0-9] ha){s<=1}'
+        self.FOREST_OPTIONS = (re.UNICODE | re.IGNORECASE)
+
+    def _extract(self, entry, extraction_results, extraction_metadata):
+        own_forest = self._find_forest(entry["text"])
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/forest_extractor.py]: `own_forest` doesn't fully convey to me what this variable exactly holds. I mean I can know by reading the implementation of `_find_forest`, but I should be able to know from the variable name what it is or holds. The more explicit, the better :relaxed:  If it's a flag for whether the person owns forest, it could be something like `owns_forest` or even `person_owns_forest`, or `has_forest_property`. If it's the area of the forest, maybe something along the lines of `owned_forest_area`. Or `forest_area_owned` :thinking:  Naming variables well is really challenging :'D

File: extractors/bookseries/karelians/extractors/kalastus_extractor.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import re
+
+import core.utils.regex_utils as regexUtils
+from core.pipeline_construction.base_extractor import BaseExtractor
+
+
+class KalastusExtractor(BaseExtractor):
+    extraction_key = 'kalastus'
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/kalastus_extractor.py]: Class name and extraction key are both Finnish. English preferred.

File: extractors/bookseries/karelians/extractors/forest_extractor.py (hunk 5 above)
Comment (lumikani) [reply extractors/bookseries/karelians/extractors/forest_extractor.py]: Generally speaking, these patterns could use some comments that explain what they are, or more importantly, why they are written the way they are. You can look at the comments for some of the other extractors.

File: extractors/bookseries/karelians/extractors/kalastus_extractor.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import core.utils.regex_utils as regexUtils
+from core.pipeline_construction.base_extractor import BaseExtractor
+
+
+class KalastusExtractor(BaseExtractor):
+    extraction_key = 'kalastus'
+
+    def __init__(self, cursor_location_depends_on=None, options=None):
+        super(KalastusExtractor, self).__init__(cursor_location_depends_on, options)
+        self.KALASTUS_PATTERN = r"(?P<kalastus>kalastus)|(kalastamista)|(kalastaa)|(kalastaminen)"
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/kalastus_extractor.py]: Code in Finnish. Also, none of these patterns allow for any typos? Did you look into how common mispellings of these words are? I have a recollection that, for example, lower case "L" often gets confused with upper case "i" or 1 by the OCR software.

File: extractors/bookseries/karelians/extractors/kalastus_extractor.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+        self.KALASTUS_OPTIONS = (re.UNICODE | re.IGNORECASE)
+
+    def _extract(self, entry, extraction_results, extraction_metadata):
+        own_kalastus = self._find_kalastus(entry["text"])
```
Comment (lumikani) [inline extractors/bookseries/karelians/extractors/kalastus_extractor.py]: Not entirely sure what this `own_kalastus` variable name is supposed to mean :sweat_smile:

**Review by lumikani — CHANGES_REQUESTED** [review]
There were some changes I want at least related to Finnish/English in code, specifically the code or the keys in the output of `kaira` should not contain any Finnish, unless absolutely unavoidable (ie. maybe something like "sisu" shouldn't be translated :'D)

I think we should set some time aside, maybe on Monday, to look into writing some more comprehensive unit tests that cover edge cases.

The religion extractor either needs some work or needs to be renamed 🤔 or maybe it could just be dropped from this completely. It was just a sample extractor and it doesn't really currently extract anything meaningful. 

The fishing, hunting and forest extractors could all be in their own commits instead of being all in one big commit. They are not really related to each other in any way - there's no reason they should be in the same commit. I find it nice when commits are as small as possible, and each commit only contains changes related to one *thing* at a time. Makes them easier to read if later on you or someone has to check the git history for what was done in this commit or something like that.

The pull request name and the branch name are not currently very descriptive. When I am looking at a list of branches or a list of pull requests, I feel that being able to determine, at a glance, what kind of changes some pull request or branch contains, is very practical and nice.

File: extractors/bookseries/karelians/extractors/forest_extractor.py (hunk 5 above)
Comment (Tumetsu) [reply extractors/bookseries/karelians/extractors/forest_extractor.py]: In addition of comment about some why some "special patterns" are used (for example the one about 'H' character in this case) it might be a good idea to add unit test which tests the particular case and in the process gives an example when the said pattern is useful/required.

File: extractors/bookseries/karelians/extractors/kalastus_extractor.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import re
+
+import core.utils.regex_utils as regexUtils
+from core.pipeline_construction.base_extractor import BaseExtractor
+
+
+class KalastusExtractor(BaseExtractor):
```
Comment (Tumetsu) [inline extractors/bookseries/karelians/extractors/kalastus_extractor.py]: When this and hunting extractors are added to the pipeline you might have to consider if the data is added to the correct person: primary person or spouse. Most of the time in the book entries the primary person is male and spouse is female but this is not always the case. 

You could do some assumptions on if the hunting and fishing activities were gendered back in the day. For example assume that most of the time fishing is done by males. You can see one example implementation of this kind of extractor here: https://github.com/Learning-from-our-past/kaira-core/blob/main/extractors/bookseries/karelians/extractors/martta_activity_flag_extractor.py 

In the above extractor we make assumption that only females were part of martta activities and the extractor activates only for the female persons. I'm not sure if you have to do 
```
if self._in_spouse_extractor and self._deps['person']['name']['gender'] == 'Male':
            is_female = True
        elif not self._in_spouse_extractor and self._deps['person']['name']['gender'] == 'Female':
            is_female = True
```
logic here though. Maybe you can find the current person's sex directly depending on current person's name property instead? Lumi can probably help with figuring that out.

Before making such assumptions, I'd suggest doing a quick search on text files searching the keywords (in this case "kalastus") and see if your assumption is reasonable. Then document the assumption to the extractor's doc-string and finally mentioning about it to John or other researchers. Or if you are unsure if the assumption makes sense, probably someone from researchers is happy to chip in with their opinion.

Also note that the male/female division is just an example of how to figure out which person the field here refers to. There might be other and better methods for that.

File: extractors/bookseries/karelians/extractors/religion_extractor.py (hunk 2 above)
Comment (Tumetsu) [reply extractors/bookseries/karelians/extractors/religion_extractor.py]: I think this could be named as `OrthodoxExtractor` or something like that. and the resulting key be `isOrthodox` or `orthodox` depending the existing convention with boolean flags.

My reasoning with using orthodox in name is that most likely everyone in the data are lutherian christians unless otherwise stated so probably we do not need a _religion extractor_ as such...? You can ask from researchers if this assumption is okay for them.

Comment (Tumetsu) [reply extractors/bookseries/karelians/extractors/religion_extractor.py]: I did a quick search to the text files with ortodoksi keyword and got varying results about people being part of some orthodoxy church activities. For example "XXX on ortodoksisen kirkon isännöitsijä" or someone being in the choir of a church. While I guess it is safe to assume them also belonging to said church, I wonder if the more accurate name for field of this data would be something like `affiliatedToOrthodoxChurch` etc? That would hopefully tell researchers more clearly that the person in question is somehow involved with the church activities and not necessarily just in the faith.

File: extractors/bookseries/karelians/extractors/tests/test_religion_extractor.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+        for text, year in expected_texts_and_years:
+            results, metadata = extractor.extract({'text': text}, {}, {})
+
+            assert results[flag] == year
+
+
+    def should_correctly_extract_religion(self, extractor):
+        self._verify_religion([
+            ('Rouva X kuuluu ortodoksisen kirkon laulu-kuoroon ja Karjalaseuraan.', True)
```
Comment (Tumetsu) [inline extractors/bookseries/karelians/extractors/tests/test_religion_extractor.py]: This example reminds me that maybe this extractor also requires some sex-detection logic so that the extracted data is assigned to the correct person? For example here the word Rouva refers clearly to the female.

I wonder if this would be a good place to experiment using conllu graphs to assign the data to the correct person? First find the keyword `ortodoksi` and then inspect the statement trying to find if there is a word indicating the identity of the person in question. In this case find out that word Rouva is related to word ortodoksi -> deduce that the female person of this text entry is in orthodoxian church.

This might well be out of scope of your current goals but might be an interesting and simple enough case to begin working with the more advanced text processing tools.

**Review by Tumetsu — CHANGES_REQUESTED** [review]
Just randomly dropped in to give some suggestions after hearing that someone is continuing this project and after checking Lumi's PR :smile:  You can discuss about my comments with Lumi if they make sense with the current state of the project and future goals since I'm mostly out of loop of the most recent developments :)

File: extractors/bookseries/karelians/extractors/tests/test_religion_extractor.py (hunk 12 above)
Comment (Tumetsu) [reply extractors/bookseries/karelians/extractors/tests/test_religion_extractor.py]: Most of the hobbies, activities etc. keywords are troublesome in how to assign that data to the correct person entry and was one of the main reasons to start using the conllu data in extraction process. It might be a good idea to develop some kind of utils for that first before going to production with extractors which do not have ability to discern the target person of the extracted data.

File: extractors/bookseries/karelians/extractors/religion_extractor.py (hunk 2 above)
Comment (lumikani) [reply extractors/bookseries/karelians/extractors/religion_extractor.py]: My initial thinking was to just drop this religion extractor, as it was just a practice/example extractor, but I like this `affiliatedToOrthodoxChurch` thinking. Maybe some researcher will get something out of just knowing that there is some link between the person and the Orthodox church. Will have to think about this a bit more :thinking: There's still the question of whether this can be implemented reliably enough without utilizing NLP, and that is still coming :sweat_smile:
